lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -265,6 +267,33 @@ void FirRegOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
     setNameFn(getResult(), getName());
 }
 
+LogicalResult FirRegOp::canonicalize(FirRegOp op, PatternRewriter &rewriter) {
+  // If the register has a symbol, we can't optimize it away.
+  if (op.getInnerSymAttr())
+    return failure();
+
+  // If the register's value is itself, we can replace the register.
+  if (op.getNext().getDefiningOp() != op)
+    return failure();
+
+  // If the register has a reset value, we can replace it with that.
+  if (auto resetValue = op.getResetValue()) {
+    rewriter.replaceOp(op, resetValue);
+    return success();
+  }
+
+  auto type = op.getType();
+
+  // Otherwise we want to replae the register with a constant 0. For now this
+  // only works with integer types.
+  auto intType = type.dyn_cast<IntegerType>();
+  if (!intType)
+    return failure();
```

Comment:
At HW/Comb level we can use `bitcast(constant(0)) or `array_create(0, 0, 0, ..., 0)`.